### ISSUE TITLE
TRIOT: iterating over (variadic !) tensors without branching

### ISSUE THREAD

**Issue by mratsim** [mratsim/Arraymancer#164]
TRIOT stands for Template Recursive Iteration Over Tensors and promises to iterate over tensors of arbitrary dimensions without branching.

Note: TRIOT as-is does not support strided iterations and must adapted for that (can it?)

# Sources
[Web page](https://alg.cs.umt.edu/triot.html)
[Stack Overflow](https://stackoverflow.com/questions/43356889/a-fast-way-to-iterate-through-multidimensional-arrays-with-different-shapes)
Paper: TRIOT: Faster tensor manipulation in C++11 [1608.00099](https://arxiv.org/pdf/1608.00099)
Git [repo](https://bitbucket.org/orserang/triot/)

# Benchmarks (Arraymancer is similar to tuple iteration)

Note: Numpy benchmark was using fixed C for loops to not penalize Python interpreter (hence should be ignored).

![image](https://user-images.githubusercontent.com/22738317/33306417-cd047e26-d412-11e7-9d87-65feb9c51afb.png)
![image](https://user-images.githubusercontent.com/22738317/33306422-d458f0da-d412-11e7-860f-ddad445f97f8.png)
![image](https://user-images.githubusercontent.com/22738317/33306455-f572a43c-d412-11e7-9bf3-8bb0271205c4.png)

# Implementation
Note, from the StackOverflow post, it seems like it can also support a variadic number of tensors

Unroll shape at compile-time
```C++
template <unsigned int DIMENSION>
inline unsigned long tuple_to_index_fixed_dimension(const_tup_t tup, const_tup_t shape) {
  unsigned long res = 0; unsigned int k;
  for (k=0; k<DIMENSION−1; ++k) {
    res += tup[k];
    res *= shape[k+1]; 
  }
  res += tup[k];
  return res; 
}

template <unsigned int DIMENSION, unsigned int CURRENT> 
class ForEachFixedDimensionHelper {
public:
  template <typename FUNCTION, typename ...TENSORS>
  inline static void apply(tup_t counter, const_tup_t shape, FUNCTION function, TENSORS & ...args) {
    for (counter[CURRENT]=0; counter[CURRENT]<shape[CURRENT]; ++counter[CURRENT]) 
      ForEachFixedDimensionHelper<DIMENSION−1, CURRENT+1>::template apply<FUNCTION, TENSORS...>(counter, shape, function, args...);
  } 
};

template <unsigned int CURRENT>
class ForEachFixedDimensionHelper<1u, CURRENT> { 
public:
  template <typename FUNCTION, typename ...TENSORS>
  inline static void apply(tup_t counter, const_tup_t shape, FUNCTION function, TENSORS & ...args) {
    for (counter[CURRENT]=0; counter[CURRENT]<shape[CURRENT]; ++counter[CURRENT]) 
      function(args[tuple_to_index_fixed_dimension<CURRENT+1>(counter, args.data_shape())]...); /* tensor.data_shape() is an accessor for returning the shape member. */
  } 
};

template <unsigned char DIMENSION> 
class ForEachFixedDimension { 
public:
  template <typename FUNCTION, typename ...TENSORS>
  inline static void apply(const_tup_t shape, FUNCTION function, TENSORS & ...args) {
    unsigned long counter[DIMENSION];
    memset(counter, 0, DIMENSION*sizeof(unsigned long)); 
    ForEachFixedDimensionHelper<DIMENSION,0>::template apply<FUNCTION, TENSORS...>(counter, shape, function, args...);
  } 
};
```

Look-up at runtime
```C++
typedef unsigned int TEMPLATE_SEARCH_INT_TYPE;
template <TEMPLATE_SEARCH_INT_TYPE MINIMUM, TEMPLATE_SEARCH_INT_TYPE MAXIMUM, template <TEMPLATE_SEARCH_INT_TYPE> class WORKER> 
class LinearTemplateSearch {
public:
  template <typename...ARG_TYPES>
  inline static void apply(TEMPLATE_SEARCH_INT_TYPE v, ARG_TYPES && ... args) {
    if (v == MINIMUM)
      WORKER<MINIMUM>::apply(std::forward<ARG_TYPES>(args)...);
    else
      LinearTemplateSearch<MINIMUM+1, MAXIMUM, WORKER>::apply(v, std::forward<ARG_TYPES>(args)...);
  } 
};

template <TEMPLATE_SEARCH_INT_TYPE MAXIMUM, template <TEMPLATE_SEARCH_INT_TYPE> class WORKER >
class LinearTemplateSearch<MAXIMUM, MAXIMUM, WORKER> { 
public:
  template <typename...ARG_TYPES>
  inline static void apply(TEMPLATE_SEARCH_INT_TYPE v, ARG_TYPES && ... args) {
    assert(v == MAXIMUM);
    WORKER<MAXIMUM>::apply(std::forward<ARG_TYPES>(args)...); 
  }
};
```

# Questions
There seems to be 2 constraints:
- DIMENSIONS, which seems similar to Arraymancer MAXRANK
- MINIMUM / MAXIMUM which would be 0 .. 2^31 - 1. The space consumed for precomputation might produce a huge binary. The linear search at compile-time might be prohibitive for fast development.

# Arraymancer implementation:

## Reference implementation
```Nim
template strided_iteration[T](t: Tensor[T], strider: IterKind): untyped =
  ## Iterate over a Tensor, displaying data as in C order, whatever the strides.

  ## Iterator init
  var coord = newSeq[int](t.rank) # Coordinates in the n-dimentional space
  var backstrides: seq[int] = @[] # Offset between end of dimension and beginning
  for i,j in zip(t.strides,t.shape):
    backstrides.add(i*(j-1))

  var iter_pos = t.offset

  ## Iterator loop
  for i in 0 ..< t.shape.product:

    ## Templating the return value
    when strider == IterKind.Values: yield t.data[iter_pos]
    elif strider == IterKind.Coord_Values: yield (coord, t.data[iter_pos])
    elif strider == IterKind.MemOffset: yield iter_pos
    elif strider == IterKind.MemOffset_Values: yield (iter_pos, t.data[iter_pos])

    ## Computing the next position
    for k in countdown(t.rank - 1,0):
      if coord[k] < t.shape[k]-1:
        coord[k] += 1
        iter_pos += t.strides[k]
        break
      else:
        coord[k] = 0
        iter_pos -= backstrides[k]
```
## in production: with init and next iterators templates

https://github.com/mratsim/Arraymancer/blob/3abe13b3a23d86536abe1151d4caba3e53f3dbe5/src/tensor/private/p_accessors.nim#L98-L131

**Comment by mratsim** [comment]
Relevant Julia fast rework in 2015: https://github.com/JuliaLang/julia/pull/10337/commits/1af945821b0dd730ab57bcdeca0d5645dba15f70

Also for parallelism on GPU we use integer div and mod which are slow, an alternative might be [int_fastdiv](https://github.com/milakov/int_fastdiv)

**Comment by mratsim** [comment]
After playing a bit with template recursion, I choose to use Nim macros instead which seems more straight forward and also do not need to search implementation at runtime.

It should be as efficient as hardcoded loop.

## Optimization
One optimization that could be added is hoisting part of the
  `realIdx = i0 * strides[0] + i1 * strides[1] + i2 * strides[2] + `... so that outer computation like `i0 * strides[0]` are not computed in the innermost loop
As is the compiler may still detect that those are invariant and do *loop hoisting optimization*.

*Pros*:
Less computation
*Cons*:
addition and multiplication are not costly, however each intermediate result `i0 * strides[0]` would need to be stored in a register, meaning 7 registers per tensor creating register pressure and if iterating on multiple tensors, register spilling.

Number of registers:
- x86: 8
- x86_64: 16
- ARM32: 14
- ARM64: 31
- MIPS: 31

## The code
```Nim
# Inspired by paper: https://arxiv.org/pdf/1608.00099
#
# Instead of using template recursion at compile-time to create nested for loops
# followed by a search at run-time that would find the appropriate
# looping sequence,
# we use the powerful Nim macros to autogenerate the nested loop statements.
# the code would be as fast as using hardcoded for loop at the price of compiling speed

# If we have an array of shape [4, 9, 7, 5] (and strides [...])
# We want to output index:
# for i in 0 ..< 4:
#   for j in 0 ..< 9:
#     for k in 0 ..< 7:
#       for l in 0 ..< 5:
#         yield i*strides[0] + j*strides[1] + ....


import ../src/tensor/backend/metadataArray
import macros

proc shape_to_strides*(shape: MetadataArray): MetadataArray {.noSideEffect.} =
  var accum = 1
  result.len = shape.len

  for i in countdown(shape.len-1,0):
    result[i] = accum
    accum *= shape[i]

const MAXRANK = 4

iterator iterShape(dim: int): int =
  # Special case if dim = 0 or 1
  # We still need to do a single iteration
  # Index will be 0 computed realIdx will be correct

  let ending = if dim == 0: 0
               else: dim-1
  for i in 0 .. ending:
    yield i

macro forParams(dim: static[int], shape, strides: MetadataArray, statement: untyped): untyped =
  result = newNimNode(nnkStmtList)

  if dim < MAXRANK:
    result.add nnkForStmt.newTree(
      newIdentNode("i" & $dim),
      nnkCall.newTree(
        newIdentNode("iterShape"),
        nnkBracketExpr.newTree(
            shape,
            newLit(dim)
        )
      ),
      getAST forParams(dim + 1, shape, strides, statement)
    )
  else:
    result.add statement

macro genNestedFor(shape, strides: MetadataArray): untyped =

  # 1. Store the loop indexes [i0, i1, i2, i3, i4, ...]
  var iters: array[MAXRANK, NimNode]
  for i in 0 ..< MAXRANK:
    iters[i] = ident("i" & $i)

  # 2. Construct the AST of the inner most statement (after the nested for loops)
  #
  #   var realIdx = 0
  #   realIdx = i0 * strides[0] + i1 * strides[1] + i2 * strides[2] ...
  #   yield realIdx
  var realIdxStmt = newNimNode(nnkStmtList)
  realIdxStmt.add(
    # var realIdx = 0
    nnkVarSection.newTree(
        nnkIdentDefs.newTree(
          newIdentNode("realIdx"),
          newEmptyNode(),
          newLit(0)
      )
    )
  )
  for i in 0 ..< MAXRANK:
    # realIdx += i0 * strides[0]
    # realIdx += i1 * strides[1]
    # ...
    realIdxStmt.add(
      newTree(
        nnkInfix,
        ident("+="),
        ident("realIdx"),
        newTree(
          nnkInfix,
          ident("*"),
          iters[i],
          newTree(
            nnkBracketExpr,
            strides,
            newLit(i)
          )
        )
      )
    )
  # yield realIdx
  realIdxStmt.add(
    nnkYieldStmt.newTree(
      ident("realIdx")
    )

  )

  # 3. Generate the nested for loop and embed the realIdx statements
  result = newNimNode(nnkStmtList)
  result.add(
    getAST forParams(
      0,
      shape,
      strides,
      realIdxStmt
    )
  )

  # TODO: currently all computations happen in the innermost loop.
  # Consideration:
  # - Part of it could be hoisted out.
  # - Given that each computation for each loop is done separately the compiler can probably
  #   detect the invariant.
  # - Hoisting the computations would mean holding the intermediate results in registers
  #   This will require 7 intermediate variables (while we already have 7 iteration variables)
  #   and might cause register pressure and register spilling

let a = [2, 3, 4].toMetadataArray
let a_st = a.shape_to_strides

echo a_st

iterator nFor(shape, strides: MetadataArray): int =
  genNestedFor(shape, strides)

for i in nFor(a, a_st):
  echo i
```

**Comment by mratsim** [comment]
Checking the ASM produce by compiling in release + march=native it shows that the Clang compiler automatically hoist the multiplications out of the inner loops:

![2018-03-14_11-04-52](https://user-images.githubusercontent.com/22738317/37395902-ba95e9b0-2777-11e8-9016-ac74ff945397.png)

I however notice a very strong impact of the maximum dimensionality on the speed of the algorithms (4 nested loops -> less than 3 sec and 7 nested loops -> 15 sec):

```Nim
var test = 0'u
for _ in 0..<100_000_000:
  for i in nFor(a, a_st):
    test += i.uint

echo test
```

**Comment by mratsim** [comment]
Probably wrong idea:
- At the (huge) price of code size: generate MAXRANK cases, so that for loops with 0 or 1 sized dimension are not used.

## Bench
[Relevant benchmarks in C/C++ of various iteration schemes](http://nadeausoftware.com/articles/2012/06/c_c_tip_how_loop_through_multi_dimensional_arrays_quickly)

![image](https://user-images.githubusercontent.com/22738317/37398247-2adbf128-277e-11e8-9572-70c4cf84ecea.png)
![image](https://user-images.githubusercontent.com/22738317/37398265-3d0cdccc-277e-11e8-8c20-ff9ef54d6675.png)

![image](https://user-images.githubusercontent.com/22738317/37398279-4564617e-277e-11e8-840a-0060897c21f4.png)
![image](https://user-images.githubusercontent.com/22738317/37398300-57baffcc-277e-11e8-925c-1fd69f7604d7.png)

**Comment by mratsim** [comment]
After implementing a full bench, it seems like nested for loops are about 30%~slower (contiguous case optimization was removed)

So the current implementation will be kept as it's faster, and probably optimal memory-wise.

Benchmark
```Nim
# Benchmark for https://github.com/mratsim/Arraymancer/issues/164
# Place it in "Arraymancer/build" folder

# Inspired by paper: https://arxiv.org/pdf/1608.00099
#
# Instead of using template recursion at compile-time to create nested for loops
# followed by a search at run-time that would find the appropriate
# looping sequence,
# we use the powerful Nim macros to autogenerate the nested loop statements.
# the code would be as fast as using hardcoded for loop at the price of compiling speed

# If we have an array of shape [4, 9, 7, 5] (and strides [...])
# We want to output index:
# for i in 0 ..< 4:
#   for j in 0 ..< 9:
#     for k in 0 ..< 7:
#       for l in 0 ..< 5:
#         yield i*strides[0] + j*strides[1] + ....


import ../src/tensor/backend/metadataArray
import macros

proc shape_to_strides*(shape: MetadataArray): MetadataArray {.noSideEffect.} =
  var accum = 1
  result.len = shape.len

  for i in countdown(shape.len-1,0):
    result[i] = accum
    accum *= shape[i]

const MAXRANK = 7

type TestTensor = object
  rank: int
  shape, strides: MetadataArray

################################################################################################
#### new nested for loop iteration scheme

iterator iterShape(dim: int): int =
  # Special case if dim = 0 or 1
  # We still need to do a single iteration
  # Index will be 0 computed realIdx will be correct

  let ending = if dim == 0: 0
               else: dim-1
  for i in 0 .. ending:
    yield i

macro forParams(dim: static[int], shape, strides: MetadataArray, statement: untyped): untyped =
  result = newNimNode(nnkStmtList)

  if dim < MAXRANK:
    result.add nnkForStmt.newTree(
      newIdentNode("i" & $dim),
      nnkCall.newTree(
        newIdentNode("iterShape"),
        nnkBracketExpr.newTree(
            shape,
            newLit(dim)
        )
      ),
      getAST forParams(dim + 1, shape, strides, statement)
    )
  else:
    result.add statement

macro genNestedFor(shape, strides: MetadataArray): untyped =

  # 1. Store the loop indexes [i0, i1, i2, i3, i4, ...]
  var iters: array[MAXRANK, NimNode]
  for i in 0 ..< MAXRANK:
    iters[i] = ident("i" & $i)

  # 2. Construct the AST of the inner most statement (after the nested for loops)
  #
  #   var realIdx = 0
  #   realIdx = i0 * strides[0] + i1 * strides[1] + i2 * strides[2] ...
  #   yield realIdx
  var realIdxStmt = newNimNode(nnkStmtList)
  realIdxStmt.add(
    # var realIdx = 0
    nnkVarSection.newTree(
        nnkIdentDefs.newTree(
          newIdentNode("realIdx"),
          newEmptyNode(),
          newLit(0)
      )
    )
  )
  for i in 0 ..< MAXRANK:
    # realIdx += i0 * strides[0]
    # realIdx += i1 * strides[1]
    # ...
    realIdxStmt.add(
      newTree(
        nnkInfix,
        ident("+="),
        ident("realIdx"),
        newTree(
          nnkInfix,
          ident("*"),
          iters[i],
          newTree(
            nnkBracketExpr,
            strides,
            newLit(i)
          )
        )
      )
    )
  # yield realIdx
  realIdxStmt.add(
    nnkYieldStmt.newTree(
      ident("realIdx")
    )

  )

  # 3. Generate the nested for loop and embed the realIdx statements
  result = newNimNode(nnkStmtList)
  result.add(
    getAST forParams(
      0,
      shape,
      strides,
      realIdxStmt
    )
  )

  # TODO: currently all computations happen in the innermost loop.
  # Consideration:
  # - Part of it could be hoisted out.
  # - Given that each computation for each loop is done separately the compiler can probably
  #   detect the invariant.
  # - Hoisting the computations would mean holding the intermediate results in registers
  #   This will require 7 intermediate variables (while we already have 7 iteration variables)
  #   and might cause register pressure and register spilling
  # - Fused multiply-add might avoid register pressure and be just as fast as add

iterator nFor_triot(t: TestTensor): int =
  genNestedFor(t.shape, t.strides)

################################################################################################
#### original iteration scheme

import ../src/tensor/backend/memory_optimization_hints

template initStridedIteration*(coord, backstrides, iter_pos: untyped, t, iter_offset, iter_size: typed): untyped =
  ## Iterator init
  var iter_pos = 0
  withMemoryOptimHints() # MAXRANK = 8, 8 ints = 64 Bytes, cache line = 64 Bytes --> profit !
  var coord {.align64, noInit.}: array[MAXRANK, int]
  var backstrides {.align64, noInit.}: array[MAXRANK, int]
  for i in 0..<t.rank:
    backstrides[i] = t.strides[i]*(t.shape[i]-1)
    coord[i] = 0

  # Calculate initial coords and iter_pos from iteration offset
  if iter_offset != 0:
    var z = 1
    for i in countdown(t.rank - 1,0):
      coord[i] = (iter_offset div z) mod t.shape[i]
      iter_pos += coord[i]*t.strides[i]
      z *= t.shape[i]

template advanceStridedIteration*(coord, backstrides, iter_pos, t, iter_offset, iter_size: typed): untyped =
  ## Computing the next position
  for k in countdown(t.rank - 1,0):
    if coord[k] < t.shape[k]-1:
      coord[k] += 1
      iter_pos += t.strides[k]
      break
    else:
      coord[k] = 0
      iter_pos -= backstrides[k]

template stridedIteration*(t, iter_offset, iter_size: typed): untyped =
  ## Iterate over a Tensor, displaying data as in C order, whatever the strides.

  initStridedIteration(coord, backstrides, iter_pos, t, iter_offset, iter_size)
  for i in iter_offset..<(iter_offset+iter_size):
    yield i
    advanceStridedIteration(coord, backstrides, iter_pos, t, iter_offset, iter_size)

iterator nFor_current(t: TestTensor): int =

  var size = 1
  for i in 0..<t.rank:
    size *= t.shape[i]
  stridedIteration(t, 0, size)


################################################################################################
#### Bench

import times


let a = [2, 3, 4].toMetadataArray
let a_st = a.shape_to_strides
let t = TestTensor(rank: 3, shape: a, strides: a_st)

proc test_triot(): uint {.noinline.}=
  for _ in 0..<100_000_000:
    for i in nFor_triot(t):
      result += i.uint

proc test_current(): uint {.noinline.}=
  for _ in 0..<100_000_000:
    for i in nFor_current(t):
      result += i.uint

##################
# Warmup
var start = cpuTime()

# Raise CPU to max perf even if using ondemand CPU governor
# (mod is a costly operation)
var foo = 123
for i in 0 ..< 100000000:
  foo += i*i mod 456
  foo = foo mod 789

var stop = cpuTime()
echo "Warmup: " & $(stop - start) & "s"
#################

start = cpuTime()
echo test_current()
stop = cpuTime()
echo "Current implementation: " & $(stop - start) & "s"

start = cpuTime()
echo test_triot()
stop = cpuTime()
echo "Macros generated nested for loop: " & $(stop - start) & "s"
```

Results on i5-5257U (dual-core mobile broadwell)
* current implementation: 4.2-4.7 seconds on 100M loops
* Fixed for loop: 5.7-6.2 seconds
![2018-03-14_13-59-48](https://user-images.githubusercontent.com/22738317/37403692-f882d608-278f-11e8-837c-6a85b6a1aaf9.png)

**Comment by mratsim** [comment]
In the quest to refactor Arraymancer backend, a properly implemented Triot has been started here: https://github.com/numforge/laser/tree/master/benchmarks/loop_iteration

Basically for doing `+=` element-wise on `x` and `y` tensor it generates the following code:

```Nim
case rank(x)
of 0:
  echo [x[[]]]
  x[[]] += y[[]]
of 1:
  for iter_dim0_267884 in 0 ..< x.shape[0]:
    x[[iter_dim0_267884]] += y[[iter_dim0_267884]]
of 2:
  for iter_dim1_268264 in 0 ..< x.shape[0]:
    for iter_dim0_268366 in 0 ..< x.shape[1]:
      x[[iter_dim1_268264, iter_dim0_268366]] += y[[iter_dim1_268264, iter_dim0_268366]]
of 3:
  for iter_dim2_268746 in 0 ..< x.shape[0]:
    for iter_dim1_268848 in 0 ..< x.shape[1]:
      for iter_dim0_268950 in 0 ..< x.shape[2]:
        x[[iter_dim2_268746, iter_dim1_268848, iter_dim0_268950]] += y[[iter_dim2_268746, iter_dim1_268848, iter_dim0_268950]]
of 4:
  for iter_dim3_269330 in 0 ..< x.shape[0]:
    for iter_dim2_269432 in 0 ..< x.shape[1]:
      for iter_dim1_269534 in 0 ..< x.shape[2]:
        for iter_dim0_269636 in 0 ..< x.shape[3]:
          x[[iter_dim3_269330, iter_dim2_269432, iter_dim1_269534, iter_dim0_269636]] += y[[iter_dim3_269330, iter_dim2_269432, iter_dim1_269534, iter_dim0_269636]]
of 5:
  for iter_dim4_270016 in 0 ..< x.shape[0]:
    for iter_dim3_270118 in 0 ..< x.shape[1]:
      for iter_dim2_270220 in 0 ..< x.shape[2]:
        for iter_dim1_270322 in 0 ..< x.shape[3]:
          for iter_dim0_270424 in 0 ..< x.shape[4]:
            x[[iter_dim4_270016, iter_dim3_270118, iter_dim2_270220, iter_dim1_270322, iter_dim0_270424]] += y[[iter_dim4_270016, iter_dim3_270118, iter_dim2_270220, iter_dim1_270322, iter_dim0_270424]]
of 6:
  for iter_dim5_270804 in 0 ..< x.shape[0]:
    for iter_dim4_270906 in 0 ..< x.shape[1]:
      for iter_dim3_271008 in 0 ..< x.shape[2]:
        for iter_dim2_271110 in 0 ..< x.shape[3]:
          for iter_dim1_271212 in 0 ..< x.shape[4]:
            for iter_dim0_271314 in 0 ..< x.shape[5]:
              x[[iter_dim5_270804, iter_dim4_270906, iter_dim3_271008, iter_dim2_271110, iter_dim1_271212, iter_dim0_271314]] += y[[iter_dim5_270804, iter_dim4_270906, iter_dim3_271008, iter_dim2_271110, iter_dim1_271212, iter_dim0_271314]]
of 7:
  for iter_dim6_271694 in 0 ..< x.shape[0]:
    for iter_dim5_271796 in 0 ..< x.shape[1]:
      for iter_dim4_271898 in 0 ..< x.shape[2]:
        for iter_dim3_272000 in 0 ..< x.shape[3]:
          for iter_dim2_272102 in 0 ..< x.shape[4]:
            for iter_dim1_272204 in 0 ..< x.shape[5]:
              for iter_dim0_272306 in 0 ..< x.shape[6]:
                x[[iter_dim6_271694, iter_dim5_271796, iter_dim4_271898, iter_dim3_272000, iter_dim2_272102, iter_dim1_272204, iter_dim0_272306]] += y[[iter_dim6_271694, iter_dim5_271796, iter_dim4_271898, iter_dim3_272000, iter_dim2_272102, iter_dim1_272204, iter_dim0_272306]]
```

Bench is using simultaneous iterations on 4 tensors.

2 kind of iterations are available:
  - `global`: iterate on the shape to get coordinate and from those index the tensors
  - `local`: Have iteration scheme per tensors at the expense of code size.

Local is much faster than global for both contiguous and non-contiguous cases.
Unfortunately TRIOT local iteration is not trivial to implement as it's not easy to yield in the middle of it, there is no natural `next`  be interleaved on multiple tensors. (coroutines ...)

Furthermore the codesize is quite big already with `global` triot, a per tensor iteration would make it explode and the macro is hard to maintain.
